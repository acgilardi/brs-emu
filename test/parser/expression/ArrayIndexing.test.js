const Parser = require("../../../lib/parser");
const BrsError = require("../../../lib/Error");
const { Lexeme, BrsTypes } = require("brs");
const { Int32 } = BrsTypes;

const { EOF } = require("../ParserTests");

describe("parser array indexing", () => {
    afterEach(() => BrsError.reset());

    test("one-dimensional indexing", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.Identifier, text: "_", line: 1 },
            { kind: Lexeme.Equal, text: "=", line: 1 },
            { kind: Lexeme.Identifier, text: "foo", line: 1 },
            { kind: Lexeme.LeftSquare, text: "[", line: 1 },
            { kind: Lexeme.Integer, text: "2", literal: new Int32(2), line: 1 },
            { kind: Lexeme.RightSquare, text: "]", line: 1 },
            EOF
        ]);

        expect(BrsError.found()).toBeFalsy();
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });

    test("multi-dimensional indexing", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.Identifier, text: "_", line: 1 },
            { kind: Lexeme.Equal, text: "=", line: 1 },
            { kind: Lexeme.Identifier, text: "foo", line: 1 },
            { kind: Lexeme.LeftSquare, text: "[", line: 1 },
            { kind: Lexeme.Integer, text: "2", literal: new Int32(2), line: 1 },
            { kind: Lexeme.RightSquare, text: "]", line: 1 },
            { kind: Lexeme.LeftSquare, text: "[", line: 1 },
            { kind: Lexeme.Integer, text: "0", literal: new Int32(0), line: 1 },
            { kind: Lexeme.RightSquare, text: "]", line: 1 },
            { kind: Lexeme.LeftSquare, text: "[", line: 1 },
            { kind: Lexeme.Integer, text: "6", literal: new Int32(6), line: 1 },
            { kind: Lexeme.RightSquare, text: "]", line: 1 },
            EOF
        ]);

        expect(BrsError.found()).toBeFalsy();
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });
});