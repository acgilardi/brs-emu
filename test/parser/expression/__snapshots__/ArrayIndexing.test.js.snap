// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser array indexing multi-dimensional indexing 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": IndexedGet {
      "closingSquare": Object {
        "kind": 3,
        "line": 1,
        "text": "]",
      },
      "index": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 6,
        },
      },
      "obj": IndexedGet {
        "closingSquare": Object {
          "kind": 3,
          "line": 1,
          "text": "]",
        },
        "index": Literal {
          "value": Int32 {
            "kind": 3,
            "value": 0,
          },
        },
        "obj": IndexedGet {
          "closingSquare": Object {
            "kind": 3,
            "line": 1,
            "text": "]",
          },
          "index": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 2,
            },
          },
          "obj": Variable {
            "name": Object {
              "kind": 21,
              "line": 1,
              "text": "foo",
            },
          },
        },
      },
    },
  },
]
`;

exports[`parser array indexing one-dimensional indexing 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": IndexedGet {
      "closingSquare": Object {
        "kind": 3,
        "line": 1,
        "text": "]",
      },
      "index": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 2,
        },
      },
      "obj": Variable {
        "name": Object {
          "kind": 21,
          "line": 1,
          "text": "foo",
        },
      },
    },
  },
]
`;
